src/libraries/Native/Unix/System.Security.Cryptography.Native/openssl.c

```diff
@@ -1282,6 +1290,9 @@ static void HandleShutdown()
 
 static int32_t EnsureOpenSsl11Initialized()
 {
+#ifdef FEATURE_DISTRO_AGNOSTIC_SSL
+    InitializeOpenSSLShim();
```

Comment:
It seems we should use the g_initLock around the InitializeOpenSSLShim here (and make its definition unconditional)

Reply:
> It seems we should use the g_initLock

Is there anything in `EnsureOpenSsl10Initialized` that is not thread-safe? 
I made it conditional in `EnsureOpenSsl10Initialized` only because there was already code preventing multiple calls. 
I do not think it is necessary though. If called twice or concurrently, would it just do exactly the same thing?

> make its definition unconditional

Do you mean remove `#ifdef FEATURE_DISTRO_AGNOSTIC_SSL` ? 
Would tat compile when we do not have the shim?

Reply:
Actually, it looks like the change just reintroduced the crash back.

We need to do initialization of the shim in `CryptoNative_EnsureOpenSslInitialized`. `EnsureOpenSsl10Initialized` is too late, by then we have crashed already while checking Ssl state.

Reply:
`CryptoNative_EnsureOpenSslInitialized` is also convenient, since we know that it will be called exactly once - from static constructor. 
Overall I now think the original location of the `InitializeOpenSSLShim` had fewer problems.

Reply:
Just noticed that we also have `CryptoNative_EnsureLibSslInitialized`, which tries to do similar initialization. It is not evidently called early enough, since we have crashes. 

The only difference is that it does `DetectCiphersuiteConfiguration` I think it should be folded into `CryptoNative_EnsureOpenSslInitialized`. 

Then we would know that it is called, and called exactly once and early enough.

Reply:
It seems that one is for openssl and the other is for other ssl (macOS)

Reply:
It looks like one calls another. However the calee (`EnsureOpenSslInitialized`) could be called on its own and would need to ensure the shim is initialized anyways.

I think getting all this initialization in a single entry point may make it easier to ensure that everything gets initialized, in the right order, and only once.